### PR TITLE
Update `CONTRIBUTING.md`

### LINKED ISSUE #1225: Update the `CONTRIBUTING.md` guide
I had added a generic contributing guide awhile back with the intentions of re-writing it and adding our project-specific bits, but unfortunately I never really got around to this.

However, now that the HAL has been unified, we are using `xtask` automation, and the examples require metadata, I think it's important that we clearly document what's required and what's expected for contributors.

Slightly tangential, but we should also update our PR template to point users to this file and ensure they have read it first.

### PR DESCRIPTION
**Pull request by playfulFence**
closes #1225 

A new contributing guide, which follows a [Mozilla guide](https://mozillascience.github.io/working-open-workshop/contributing/) (or at least is trying to)

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -1,101 +1,92 @@
...
+*   We may ask for changes to be made before a PR can be merged, either using [suggested changes](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/incorporating-feedback-in-your-pull-request) or pull request comments. You can apply suggested changes directly through the UI. You can make any other changes in your fork, then commit them to your branch.
+* As you update your PR and apply changes, mark each conversation as [resolved](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#resolving-conversations).
+*   Resolve merge conflicts if they arise, using resources like [this git tutorial](https://github.com/skills/resolve-merge-conflicts) for help.
 
-[link PR to issue]: https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue
```
[the remark this answers is not in the document]
Comment (playfulFence) [reply CONTRIBUTING.md:93]: Yes, totally agree, separate links look much better in raw MD

Will do the fix

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -1,101 +1,92 @@
...
+Quick Navigation
+----------------
 
-To get an overview of the project, please read the [README]. Here are some resources to help you get started with open source contributions:
+*   [New Contributor Guide](#new-contributor-guide)
+*   [Getting Started](#getting-started)
+    *   [Issues: Reporting and Resolving](#issues-reporting-and-resolving)
+    *   [Making Changes: Fork, Edit, and Pull Request](#making-changes-fork-edit-and-pull-request)
+*   [Your Pull Request: From Submission to Merge](#your-pull-request-from-submission-to-merge)
+*   [Community and Communication](#community-and-communication)
```
[the remark this answers is not in the document]
Comment (playfulFence) [reply CONTRIBUTING.md:17]: IMO: 
1) having `Quick Navigation` part in text is a bit more reader-friendly and beautiful 😄 
2) "Outline" function is rarely used, from my experience, but of course feel 

However, if you insist and think it's a critical point - will revert it back

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -1,101 +1,92 @@
...
 
-### Issues
+Getting Started
+---------------
 
-#### Create a New Issue
+### Issues: Reporting and Resolving
 
-If you spot a problem with the docs, [search if an issue already exists]. If a related issue doesn't exist, you can open a new issue using the [issue form].
+#### Report a New Issue
```
[the remark this answers is not in the document]
Comment (playfulFence) [reply CONTRIBUTING.md:35]: Agree, wbd

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -1,101 +1,92 @@
...
 
-[search if an issue already exists]: https://docs.github.com/en/github/searching-for-information-on-github/searching-on-github/searching-issues-and-pull-requests#search-by-the-title-body-or-comments
-[issue form]: https://github.com/esp-rs/esp-hal/issues/new/
+Encountered a problem or have an idea? First, [check existing issues](https://github.com/esp-rs/esp-hal/issues) to avoid duplicates. If your concern is new, use our [issue form](https://github.com/esp-rs/esp-hal/issues/new/) to submit it.
 
-#### Solve an Issue
+#### Work on an Issue
```
[the remark this answers is not in the document]
Comment (playfulFence) [reply CONTRIBUTING.md:39]: Agree, wbd

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -1,101 +1,125 @@
-# Welcome to the `esp-hal` contributing guide
 
-Thank you for investing your time in contributing to our project!
+Welcome to the `esp-hal` Contributing Guide
+===========================================
```
Comment (jessebraham) [inline CONTRIBUTING.md:3]: Can we please use hashes for headers, all of our other `README.md` files use this convention.

Comment (playfulFence) [reply CONTRIBUTING.md:3]: Oops, sorry, bad habit, will be fixed

**Comment by jessebraham** [comment]
(Sorry I'm reading some material relating to contributing guides, will do a proper review within the next few business days)

**Comment by jessebraham** [comment]
Sorry for not getting to this, had a lot on my plate.

I think we should meet as a team to discuss how best to move forward with this. I think there is a fair bit of information that should ultimately be included in `CONTRBUTING.md` (and/or other documents), so it'd be nice to come up with a bit of a game plan first.

CC @MabezDev

**Comment by MabezDev** [comment]
Now that we have the xtask, I think we should also document some do's and don't of how to use it. For example we should state that all xtask commands should be run from the root of the project, not in a sub project.

**Comment by playfulFence** [comment]
@MabezDev, agree!
Maybe let's discuss this topic on a next meeting(/when most of us won't have a vacation 😄)? Just to put this matter to rest.

cc: @jessebraham (as you've had some thoughts and suggestions on this)

**Review by jessebraham — APPROVED** [review]
Sorry that this took so long, but I think this looks fine and we should just merge it. Thanks for your patience!

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -1,70 +1,100 @@
...
 
-We ask that you ensure all source code files has been properly formatted with `rustfmt`, and that you have linted your changes by running `cargo clippy`. These tools can be installed by running the following commands:
+### Testing Your Contributions
+
+Ensuring the quality and reliability of `esp-hal` is a shared responsibility, and testing plays a critical role in this process. Our GitHub CI automatically checks the buildability of all examples and drivers within the project. However, automated tests can't catch everything, especially when it comes to the nuanced behavior of hardware interactions. So make sure that the example affected by your change works as expected. 
+
+Further steps that can (or should) be taken in testing: 
+
+* Using [xtask], build examples for the specified chip. 
+* Build the documentation if it has been modified using the `run-doc-test` command in [xtask].
```
Comment (SergioGasquez) [inline CONTRIBUTING.md:90]:
```suggestion
* Build the documentation and run the doctests if they have been modified using the `build-documentation` and  `run-doc-test` commands in [xtask].
```

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -1,70 +1,100 @@
...
+* Run the [HIL] tests locally if changes have been made to them. 
+
+[xtask]: https://github.com/esp-rs/esp-hal/tree/main/xtask
+
+### Commit Your Updates
+
+Commit your changes once you're satisfied. Review your own work to streamline the review process later. Use `rustfmt` and `cargo clippy` to ensure your code adheres to Rust's conventions.
```
Comment (SergioGasquez) [inline CONTRIBUTING.md:97]: We should suggest using `lint-packages` and `fmt-packages` `xtask `subcommands

**Review by SergioGasquez — APPROVED** [review]
LGTM! Just a pair of comments of using more xtask subcommands

File: CONTRIBUTING.md (hunk 7 above)
Comment (playfulFence) [reply CONTRIBUTING.md:97]: `fmt-packages` is already suggested (for some reason I posted a while back that it was `fmt-workspace`, so thanks for the comment, typo corrected) 

`lint-packages` suggestion was also added

**Review by JurajSadel — APPROVED** [review]
Thanks!

**Review by MabezDev — APPROVED** [review]

**Pull request merged**